src/Avalonia.X11/X11Window.cs

```diff
@@ -1146,6 +1150,35 @@ public void SetIcon(IWindowIconImpl icon)
             }
         }
 
+        /// <summary>
+        ///   Sets the application name that is visible for example in the Gnome desktop top bar
+        /// </summary>
+        /// <param name="name">The application name to set</param>
+        public void SetApplicationName(string name)
+        {
+            var classHint = XAllocClassHint();
+
+            if (classHint == IntPtr.Zero)
+            {
+                return;
+            }
+
+            var classHintData = Marshal.PtrToStructure<XClassHint>(classHint);
+
+            var classTextData = Encoding.ASCII.GetBytes(name);
+            fixed (void* textPtr = classTextData)
+            {
+                classHintData.ResClass = new IntPtr(textPtr);
+                classHintData.ResName = new IntPtr(textPtr);
```

Comment:
According to https://tronche.com/gui/x/icccm/sec-4.html#WM_CLASS the instance name should be taken from `RESOURCE_NAME` or match the executable file name.

Also note that human-readable application-title is supposed to be specified in the .desktop-file shipped with your application.

Reply:
I re-read the documentation I was going off on: https://linux.die.net/man/3/xclasshint and sure enough it does also mention that WM_CLASS info. If I change the code here to use that value, then it is no longer user controllable with the "Name" property in the Application XAML, which is where the application name property is so people would be stuck with their .csproj name being the name showing up for their application in a user readable place. That's not exactly a kind of limitation I'd want to have. It would also be need to documented that the Name property in Application doesn't work on Linux.

Reply:
Reading that documentation again it looks like maybe after all this existing code:
```
                XChangeProperty(_x11.Display, _handle, _x11.Atoms.XA_WM_CLASS, _x11.Atoms.XA_STRING, 8,
                    PropertyMode.Replace, pdata, data.Length);
```
is incorrect as it provides only one string and not two? It seems in the documentation you linked the second string in WM_CLASS is the name of the application like "Emacs" and as Avalonia doesn't set that currently, seeing it be empty makes sense to me.

Reply:
See https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html and https://specifications.freedesktop.org/startup-notification-spec/startup-notification-0.1.txt 

GNOME-recognized application name is supposed to come from an installed `.desktop` file.

Which is why initial code uses a special property in `X11PlatformOptions`. We could also fall back to `Application.Current.Name` if it's not set, but I'd rather avoid using a human-readable name as a class name.

Reply:
I think I may be onto something with specifying null terminator separated data in WM_CLASS. If I have code like:
`Encoding.ASCII.GetBytes($"first part\0{_platform.Options.WmClass}")` the "first part" shows up in the gnome top bar, and when I check the xproperties of the window:
```
$ xprop WM_CLASS
WM_CLASS(STRING) = "Avalonia Application", "first part"
```
Assuming the first part there is the internal class name, which is now what you said it should be.
Maybe that is the right way to do it?

Reply:
https://tronche.com/gui/x/icccm/sec-4.html#WM_CLASS

>- A string that names **the particular instance** of the application to which the client that owns this window belongs. Resources that are specified by instance name override any resources that are specified by class name. Instance names can be specified by the user in an operating-system specific manner. On POSIX-conformant systems, the following conventions are used:
>---        If "-name NAME" is given on the command line, NAME is used as the instance name.
>---        Otherwise, if the environment variable RESOURCE_NAME is set, its value will be used as the instance name.
>---        **Otherwise, the trailing part of the name used to invoke the program (argv[0] stripped of any directory names) is used as the instance name.** 
>- A string that names **the general class of applications to which the client that owns this window belongs**. Resources that are specified by class apply to all applications that have the same class name. **Class names are specified by the application writer**. Examples of commonly used class names include: "Emacs", "XTerm", "XClock", "XLoad", and so on.

So only `res_class` string should be application-controllable, `res_name` should be automatically detected.

The class name affects `.desktop` file search, so it should **not** contain a human-readable application name.
e. g. for GNOME Terminal `WM_CLASS(STRING) = "gnome-terminal-server", "Gnome-terminal"`. 
while .desktop file contains those lines with localized application names and the class:

```
StartupWMClass=Gnome-terminal
Name=Terminal
Name[uk]=Термінал
Name[vi]=Thiết bị cuối
Name[wa]=Terminå
Name[xh]=i-Terminal
Name[zh_CN]=终端
Name[zh_HK]=終端機
Name[zh_TW]=終端機
```

Reply:
Is your suggestion then that the code is right, but documentation is not? As in the documentation doesn't mention that on Linux you must have a .desktop file installed with your Avalonia program and you can't make a freestanding program that behaves properly? I'm more interested in making a freestanding program that could be used by just unzipping and double clicking it instead of needing an install script or something to setup a .desktop file (AFAIK they can't have relative paths in the Exec property)
Or is there some way that both situations could work?

Reply:
>Is your suggestion then that the code is right, but documentation is not? 

The current code in `master` doesn't set the second string, so WM_CLASS value is technically incorrect. However WM_CLASS **should not** contain a human-readable application name.

>on Linux you must have a .desktop file installed

On Linux _running GNOME_ desktop environment. It's their choice to ignore `WM_NAME` and `_NET_WM_NAME` properties in favor of .desktop files, not ours. So we should conform to the spec and DE guidelines in the same way as built-in GNOME apps do.

Reply:
See https://wiki.gnome.org/Projects/GnomeShell/ApplicationBased

Reply:
I get that, but all other GUI frameworks I've ever used put *something* there in the top bar by default. 
I just tested and it is true that having a .desktop file makes things work without my changes.
I then tested one more thing, if I undo my other changes but modify `SetWmClass` to have the class string like this:
```c#
var data = Encoding.ASCII.GetBytes($"{wmClass}\0{wmClass}");
```
then it seems like the default name will be shown when running the program by itself but with an installed .desktop file the name from there gets used. If I replace all my changes in this PR with just that one line change, would that be acceptable? That way both the case of not having a .desktop file kind of works, and the case of having a .desktop file looks to work as expected.

Reply:
Here are WM_CLASS values for various apps from my machine built with different UI toolkits:
"jetbrains-rider", "jetbrains-rider"
"gnome-terminal-server", "Gnome-terminal"
"Telegram", "TelegramDesktop"
"Navigator", "Firefox-esr"
"google-chrome", "Google-chrome"

As you can see, none of them uses human-readable string for WM_CLASS values.

We need to investigate how those frameworks you are referring to are providing GNOME with a meaningful application name, because specifying it as WM_CLASS definitely isn't the way to do it.

Reply:
Also note, that WM_CLASS can not be used for human-readable names because  [according to the Xlib docs](https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/XSetClassHint.html) the property has to be encoded with Host Portable Character Encoding which can only contain Latin-1 character set.

Reply:
I suspect those are installed programs .desktop files? If so they don't really apply here as there you can do whatever because it looks like it just overwrites anything the program defines with what's in the .desktop file (https://github.com/AvaloniaUI/Avalonia/pull/8595#discussion_r929677638)

Godot engine works perfectly fine without any .desktop files and it looks like they absolutely don't care and instead set the second part to be what they want to show to the user:
```
WM_CLASS(STRING) = "Godot_Engine", "Godot Third-Person Shooter Demo"
```

Also at this point I'm not even suggesting anymore that Avalonia would put a random value there, I'd be perfectly happy by having Avalonia set *both* parts of the WM_CLASS and not just one, to the same value that is currently set.